Add support for User Impersonation

Ralnoc opened the pull request:
Add ability to declare principle username to support User Impersonation within Presto. This allows the authentication with an user configured for supporting username assumption so that all queries run as the said user.

bkyryliuk commented on pyhive/presto.py:
```diff
@@ -80,7 +80,7 @@ class Cursor(common.DBAPICursor):
     """
 
-    def __init__(self, host, port='8080', username=None, catalog='hive',
+    def __init__(self, host, port='8080', username=None, principle_username=None, catalog='hive',
```
could you please provide a bit more context, maybe in the comment on differences between username & principle_username and share more information why username cannot be used ?

Ralnoc replied:
Comment added detailing the difference between the two fields and also added a documentation link to the feature in question.

Ralnoc commented:
Absolutely! I actually am going to review my changes as well to make sure no additional work is needed.

The specific use case for this scenario is that WebUIs which leverage pyhive to interact with Presto are more often using SAML for authentication. Because of this, they have a username for the user, but no password because that is not communicated in the SAML Assertion payload. In that situation, when securing Presto they configure a service account to authenticate against Presto with. 

Unfortunately this means all security is applied based on that service account, and prevents you from creating user specific permissions when accessing data or user specific resource groups for controlling how much cluster resources they can leverage.

With this change in place, it allows you to authenticate with the service account that is passed through by the application, and then tell Presto to execute the query as if it is run by the principle username. As long as Presto has been configured to allow the service account to execute queries as the principle username, presto will do so, applying the specific permissions and resource groups that are assign to that user.

This allows you to leverage a user based resource allocation for the cluster and a user based permission model on the data being accessed through Presto.

elukey commented:
I have a similar use case, really interested in this pr :)

Ralnoc commented:
@bkyryliuk -
I've confirmed that this code change should cover the specific use case and updated the docstring a bit to be more accurate.

In addition, here is the Presto documentation as it relates to this feature: https://docs.starburstdata.com/latest/security/impersonation.html

 Please let me know if you would like me to add any more information to the PR.

elukey commented:
@bkyryliuk the CI seems having problems with Python 2.7, not sure why, but the pull request seems good.

bkyryliuk commented:
@elukey @Ralnoc  - could you rebase the diff ? Master is green.
I am very not familiar with this flow and will land it once the build is fixed.
However could you explain here or in code why it is not possible to use username for this use case?
Looking into the implementation of this feature it seems to me that you can just set username & achieve the same goals. Probably I am missing smth obvious.

Ralnoc commented:
@bkyryliuk -
I've rebased on master which appears to have resolved the testing issues and added a comment to the code that details the need and use case for this change. Please let me know if you need more information.

bkyryliuk commented:
@Ralnoc this is great! Thank you

Ralnoc commented:
Glad to help! Going to dig into using this in a few places as soon as the next version releases.

tooptoop4 commented:
@Ralnoc do u use superset? i'm not sure what https://github.com/apache/incubator-superset/pull/3404 does if your PR was only just merged now.   i'm trying to solve https://github.com/apache/incubator-superset/issues/7170

Ralnoc commented:
@tooptoop4 -

I believe Superset uses SQLAlchemy to interact with Presto, instead of PyHive. That being said, reading over that issue in Superset, I would suggest you make sure that you have the Impersonation Rules defined in Presto. https://docs.starburstdata.com/latest/security/impersonation.html#user-translation-rules Without those rules in place, I believe you will get a 401 error as the attempt to execute a query as another user is denied.
